state/protocol/badger/snapshot.go

```diff
@@ -295,12 +295,49 @@ func (s *Snapshot) Pending() ([]flow.Identifier, error) {
 	return s.pending(s.blockID)
 }
 
-func (s *Snapshot) pending(blockID flow.Identifier) ([]flow.Identifier, error) {
+func (s *Snapshot) ValidPending() ([]flow.Identifier, error) {
+	pendingIDs, err := s.lookupChildren(s.blockID)
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup children for %v: %w", s.blockID, err)
+	}
+
+	return s.validPending(pendingIDs)
+}
+
+func (s *Snapshot) lookupChildren(blockID flow.Identifier) ([]flow.Identifier, error) {
 	var pendingIDs []flow.Identifier
 	err := s.state.db.View(procedure.LookupBlockChildren(blockID, &pendingIDs))
 	if err != nil {
 		return nil, fmt.Errorf("could not get pending children: %w", err)
 	}
+	return pendingIDs, nil
+}
+
+func (s *Snapshot) validPending(pendingIDs []flow.Identifier) ([]flow.Identifier, error) {
+	var validPendingIds []flow.Identifier
+	for _, pendingID := range pendingIDs {
+		children, err := s.lookupChildren(pendingID)
+		if err != nil {
+			return nil, fmt.Errorf("could not get pending children for block (%v): %w", pendingID, err)
+		}
+		// valid pending are unfinalized blocks that have at least one children
```

Comment:
This logic equally applies to finalized blocks (we just don't use it in this context yet).